cache/cache.go

```diff
@@ -289,3 +303,92 @@ func (p2 *progressWriterAt) WriteAt(p []byte, off int64) (n int, err error) {
 	p2.task.Add(n)
 	return n, errors.WithStack(err)
 }
+
+// matches: https://github.com/{OWNER}/{REPO}/releases/download/{TAG}/{ASSET}
+var githubRe = regexp.MustCompile(`^https\://github.com/([^/]+)/([^/]+)/releases/download/([^/]+)/([^/]+)$`)
+
+type githubReleaseInfo struct {
+	owner, repo, tag, asset, token string
+}
+
+func getGithubReleaseInfo(uri string) (*githubReleaseInfo, bool) {
+	g := &githubReleaseInfo{}
+	g.token = os.Getenv("HERMIT_GITHUB_TOKEN")
+	if g.token == "" {
+		return nil, false
+	}
+	m := githubRe.FindStringSubmatch(uri)
+	if len(m) != 5 {
+		return nil, false
+	}
+	var err error
+	if g.owner, err = url.PathUnescape(m[1]); err != nil {
+		return nil, false
+	}
+	if g.repo, err = url.PathUnescape(m[2]); err != nil {
+		return nil, false
+	}
+	if g.tag, err = url.PathUnescape(m[3]); err != nil {
+		return nil, false
+	}
+	if g.asset, err = url.PathUnescape(m[4]); err != nil {
+		return nil, false
+	}
+	return g, true
+}
+
+// GHRelease is a minimal type for GitHub releases meta information
+// retrieved via the GitHub API see:
+// https://docs.github.com/en/rest/reference/repos#list-releases
+type GHRelease struct {
```

Comment:
Make this private?

Reply:
oh, I thought this had to be public for JSON unmarshalling. TIL.